ImageView improvements

johnhaddon opened the pull request:
This rewrites most of the ImageView code, improving performance significantly and fixing several important bugs. The redesign follows the pattern established for the SceneView :

- The meat of the drawing is refactored into a reusable ImageGadget class, which manages a cache of texture tiles it uses for rendering.
- The ImageView defers to the ImageGadget to do most of the work.
- Additional features like the colour inspector are implemented as bolt-on classes which add functionality to the view in a non-intrusive manner.

Over time I expect the bolt-on classes may get refactored out into the public API, so that different applications can build viewers with different customisations via config files.

There is some functionality that hasn't made it through the rewrite - the selection of areas for use with an ImageStats node, and the overlay for the colour inspector is now at the top, where there isn't really room for displaying such stats. What I'd like to do is work on the NodeToolbar and PlugLayout classes to add the concept of different toolbars for the top/bottom/left/right of the viewer, and then move the inspector to the lower toolbar. I'm happy to do that next or to defer it till later - mainly I felt that this work was dragging on a bit too long, and I should at least get something visible now.

andrewkaufman commented:
I agree there isn't really room for those stats at the top, and its definitely distracting to have them there... X,Y in particular which updates even when outside the dataWindow (that was happening before as well, but I guess it was less noticable at the bottom in a black bar).

Also, most of the other toolbar widgets seem to be pushed out of the pane unless I fullscreen the Gaffer window...

johnhaddon commented:
So I should do the stuff for toolbars on all sides then?

andrewkaufman commented:
Yeah, I think so. I also wish that framing (F) wouldn't have the overlay on top of the image.... I basically always frame and then zoom out again.

johnhaddon commented:
Yeah, same here. The easy fix is to not use overlays so that the GL widget stops where the toolbar begins, but having gone to the trouble of making overlays work, I'm a bit reluctant to ditch them. Maybe we need some sort of "safe area" concept for the viewport?

andrewkaufman commented:
Yeah, that make sense

andrewkaufman commented:
I'm also seeing some weird behavior, though I can't easily reproduce it. I was sliding the whitepoint.r on a grade for a particular input image, and I got a sudden jump in behaviour when I go from 0.9 to 0.89. When I try it in a fresh session its smooth though. So maybe I hit a bug in the first session and now those results are cached? Though enabling/disabling the cache doesn't seem to help...

![grade 9](https://cloud.githubusercontent.com/assets/3979682/11788913/43d79ce4-a248-11e5-8975-911003b2b859.png)

![grade 89](https://cloud.githubusercontent.com/assets/3979682/11788922/4c5cc376-a248-11e5-88b2-e24078aedebb.png)

johnhaddon commented:
Could you mail me the node network for that one? There's a couple of possibilities, one of which is that because the viewer is now smarter, it's exposing bugs in the hashing/dirtypropagation that the dumber viewer was immune to.

andrewkaufman commented:
Sent

johnhaddon commented:
I've updated this PR with the stuff for doing toolbars on all sides, and moved the colour inspector to the bottom. I haven't been able to reproduce the grade problem you saw though, and also I couldn't find an email with the original problem script in it - am I being daft?

andrewkaufman commented:
Can you remind me where the new commits start?

Also, lets ignore that crazy thing I saw unless I manage to make it happen again... I've still only had it happen in the one session. Even the same file in a new session didn't have that problem...

johnhaddon commented:
Where the PlugLayout changes start, at a48e70e.

andrewkaufman commented:
I like the look of the new bottom overlay. I'm not too keen on the drop shadow on the top bits, but I'm not against it enough to argue...

andrewkaufman commented:
I guess we still don't have "safe area" concept for the viewport, so I still frame and then immediately zoom out. Thoughts on adding that?

andrewkaufman commented:
When I read a missing image, I get a ton of errors and then a viewer that looks like this:

![badimage](https://cloud.githubusercontent.com/assets/3979682/11997339/e2b20c18-aa2d-11e5-8843-0c52ef795226.png)

johnhaddon commented:
Updated with the following :

- Additional commits to handle errors in ImageView. The errors are just ignored, leaving it to the node graph to display them - this matches the error handling we do in SceneView. Now we have the ability to have multiple overlays, it might be nice if the Viewer had a Node::errorSignal() handler that popped up an error overlay for all view types, but I've left that for the future.
- A rebase to include the binary compatibility note for ImageView::update().

The rebase has blown away a comment I added to explain the rootSection vs rootSectionName naming - it is simply to match the existing naming of the section arguments elsewhere. I felt that keeping the publicly visible names consistent trumped the internal naming.

As for the "safe area" idea, as you say that is not introduced in this PR. I do expect we'll introduce it in the future, but I felt that this PR had already achieved its initial aims (fix all the bugs, speed things up significantly), and taken far longer than anticipated already.

andrewkaufman commented:
Looks good to me. I can't tell why codact is complaining though... just because the complexity went up?

johnhaddon commented:
Yeah, I'm a bit miffed at Codacy's assessment of my coding! The duplication it says it detected in Viewer.py isn't actually duplication - just the declaration of two similar widgets. At one point I coded up all that widget creation to be driven by a little data table to avoid duplication, but to my mind that turned out harder to follow so I went back to what we have now. I guess the complexity of some of the other UI files has increased, but I didn't think in a disproportionate way, and don't really see how else to approach it (and Codacy has no useful suggestions). The other thing is that Codacy is blind to C++, so hasn't seen the big reduction in complexity within ImageView.cpp, which was one of the main motivators for this PR.

andrewkaufman commented:
Is there any way to tell it we don't care about complexity increases? It seems like an odd thing to consider a failing test... if we're adding functionality, then its pretty likely for complexity to increase regularly...

johnhaddon commented:
I can't see a way of turning it off unfortunately - I'm kindof interested to leave it on for a bit anyway, to see if it feels useful at any point. It's definitely doing some daft stuff though, like penalising the addition of ImageGadgetTest.py with a +1 for complexity, when it's a really straightforward file, and we want all the test coverage we can get. Let's see how it goes for a bit and I'll raise it with their support if we feel like it's not useful...
